### PR TITLE
Fixes #3260 - Allows puppet to manage value of smart class parameter that can be overrided

### PR DESCRIPTION
**Pull request by orrabin**

### CODE REVIEW

**Comment by orrabin** [comment]
[test]

**Comment by orrabin** [comment]
[test]

File: app/services/classification/class_param.rb [hunk 1]
```diff
@@ -10,7 +10,11 @@ def enc
         klasses[klass.name] ||= {}
         if key_hash[klass.id]
           key_hash[klass.id].each do |key|
-            klasses[klass.name][key.to_s] = value_of_key(key, values)
+            value_of_key = value_of_key(key, values)
```
Comment (mbacovsky) [inline app/services/classification/class_param.rb]: Is there reason to have same name for the var and the method?

Comment (orrabin) [inline app/services/classification/class_param.rb]: no, will change it to `key_value` later

File: app/services/classification/base.rb [hunk 2]
```diff
@@ -57,11 +57,11 @@ def values_hash options={}
...
       value = if values[key.id] and values[key.id][key.to_s]
-                values[key.id][key.to_s][:value]
+                {:value => values[key.id][key.to_s][:value], :managed => values[key.id][key.to_s][:managed]}
               else
-                key.default_value
+                {:value => key.default_value, :managed => key.default_managed_by_puppet}
               end
-      @safe_render.parse(value)
+      @safe_render.parse(value[:value]) unless value[:managed]
     end
```
Comment (mbacovsky) [inline app/services/classification/base.rb:73]: value_of_key returns nil when value[:managed] is true. I'd find it more readable if that is  explicit. Other option is to set value to nil  in `values_hash` method and not keep track of whether it is managed by puppet in the hash. What do you think?

**Comment by mbacovsky** [comment]
I'd find it useful if user can set managed by puppet in parameters in hostgroup/host form.

**Comment by mbacovsky** [comment]
I tested UI and API and it works smoothly.
Not apparent why the CI tests failed though...

**Comment by orrabin** [comment]
@mbacovsky thanks, the CI failed fir everything yesterday. Dominic fixed the tests so now [test] should be fine

File: app/services/classification/base.rb (hunk 2 above)
Comment (orrabin) [inline app/services/classification/base.rb:73]: I use the nil to check if it should be skipped, I'll change this to be explicit

**Comment by orrabin** [comment]
@mbacovsky I've rebased this and added the option to set `managed by puppet` in host and hostgroup form

**Comment by ohadlevy** [comment]
@orrabin I've made minor modifications at https://gist.github.com/anonymous/13e402b93362ae27ce69, would you mind having a look?
thanks

File: app/views/common_parameters/_inherited_parameters.html.erb [hunk 3]
```diff
@@ -3,22 +3,25 @@
...
-  <% keys.each do |name| %>
-      <tr>
-        <%= "<td class='col-md-3' rowspan='#{keys.size}'>#{_('Global')}</td>".html_safe if name == keys.first%>
-        <td class='col-md-2'><%= content_tag :span, name, :id=>"name_#{name}", :class => "col-md-2" %></td>
-        <td class='col-md-7'><%= parameter_value_field inherited_parameters[name] %></td>
-        <td>
-        <%= link_to_function(_("override"), "override_param(this)", :title => _("Override this value"),
-                             :'data-tag' => 'override', :class =>"btn btn-default" ) if authorized_via_my_scope("host_editing", "create_params") %>
-      </tr>
+  <% if inherited_parameters.present? %>
```
Comment (ohadlevy) [inline app/views/common_parameters/_inherited_parameters.html.erb:11]: I'm missing some user guidance here, can you add some popover with guidance text.
I would try a simple text like, revert value to its puppet default value (or something on those lines).

**Comment by orrabin** [comment]
@ohadlevy I fixed your comments and added guidance in the host/hostgroup form.

File: app/assets/javascripts/lookup_keys.js [hunk 4]
```diff
@@ -147,17 +147,20 @@ function toggleOverrideValue(item) {
...
   type_field.attr('disabled', override ? null : 'disabled');
   validator_type_field.attr('disabled', override ? null : 'disabled');
-  default_value_field.attr('disabled', override ? null : 'disabled' );
+  default_value_field.attr('disabled', override && !$(default_managed_by_puppet_field).is(':checked') ? null : 'disabled' );
+  default_managed_by_puppet_field.attr('disabled', override ? null : 'disabled' );
   pill_icon.attr("class", override ? 'glyphicon glyphicon-flag' : "glyphicon- ");
   override_value_div.toggle(override);
 }
 
+
```
Comment (ohadlevy) [inline app/assets/javascripts/lookup_keys.js]: remove please.

**Comment by ohadlevy** [comment]
@orrabin +1 for code, just please amend the description text / managed by
puppet as discussed on irc.

**Comment by orrabin** [comment]
@ohadlevy I changed the label and explanation text according to @GregSutcliffe suggestions

**Comment by ohadlevy** [comment]
merged as 8e53b7c thanks @orrabin

**Pull request closed without merging**